File: day-09/encoding_error.go
package main

import (
	"fmt"
	"io/ioutil"
	"sort"
	"strconv"
	"strings"
)

const input_file string = "./day-09/input.txt"

func GetInts(input string) []int {
	var numbers []int
	split := strings.Split(strings.TrimSpace(input), "\n")
	for _, num := range split {
		float, _ := strconv.Atoi(num)
		numbers = append(numbers, float)
	}
	return numbers
}

func FindFirstNonSum(numbers []int, window int) int {
	for i := window; i < len(numbers); i++ {
		foundSum := false
		for j := i - window; j < i; j++ {
			for k := j + 1; k < i; k++ {
				if numbers[j]+numbers[k] == numbers[i] {
					foundSum = true
				}
			}
		}
		if !foundSum {
			return numbers[i]
		}
	}
	return -1
}

func Part1(input string, window int) int {
	entries := GetInts(input)
	return FindFirstNonSum(entries, window)
}

func Part2(input string, window int) int {
	entries := GetInts(input)
	target := FindFirstNonSum(entries, window)

	low := 0
	high := 1
	for high < len(entries) {
		sum := 0
		for _, v := range entries[low : high+1] {
			sum += v
		}
		if sum == target {
			sort.Ints(entries[low : high+1])
			return entries[low] + entries[high]
		} else if sum < target {
			high++
		} else {
			low++
		}
	}

	return -1
}

func main() {
	b, err := ioutil.ReadFile(input_file)
	if err != nil {
		fmt.Print(err)
	}
	data := string(b)

	fmt.Printf("Part 1: %d\n", Part1(data, 25))
	fmt.Printf("Part 2: %d\n", Part2(data, 25))
}

File: day-09/encoding_error.py
import itertools

input_file = 'day-09/input.txt'

def find_first_non_sum(numbers, window):
  for i in range(window, len(numbers)):
    window_combs = [combo for combo in itertools.combinations(numbers[i-window:i], 2) if sum(combo) == numbers[i]]
    if len(window_combs) == 0:
      return numbers[i]

def part1(input, window = 25):
  numbers = list(map(int, input.splitlines()))
  return find_first_non_sum(numbers, window)

def part2(input, window = 25):
  numbers = list(map(int, input.splitlines()))
  target = find_first_non_sum(numbers, window)

# Sliding window
  low = 0
  high = 1
  while high < len(numbers): 
    sum_size = sum(numbers[low:high+1])
    if sum_size == target:
      return min(numbers[low:high+1]) + max(numbers[low:high+1])
    elif sum_size < target:
      high += 1
    else:
      low += 1

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))